### PR TITLE
DAOS-5554 dtx: async batched commit distributed transactions

### PR DESCRIPTION
**Pull request by Nasf-Fan**
Originally, for the distributed transaction that cross multiple
redundancy groups, we will sychronously commit it when all the
participants 'prepared'. Such sync commit may introduce latency
for related distributed transaction.

This patch adjusts the commit behavior, similar as the case of
stand-alone modification, the 'committable' DTX will be cached
on the leader, and will be batched committed via dedicated ULT
sometime later. It is expected that some transactions may have
some participants on the same DAOS target(s), then the batched
commit can save some DTX commit RPCs.

Under such async batched commit mode, the DTX status on leader
may be 'committable', but is 'prepared' on non-leader. If some
read request hits these 'prepared' DTX on non-leader (in VOS),
then related caller (object) will get -DER_INPROGRESS together
with related DTX information (ID, leader), and then the caller
will query with related DTX leader whether it is committable
or not. If yes, reply the non-leader and then commit such DRX
to avoid further check by others. Introduce 'DTX_REFRESH' RPC
for such purpose.

Another main change is for DTX resync to handle the distributed
transaction that lost some whole redundancy group. Under such
case, the DTX resync logic cannot make the decision whether
commit or abort related transaction, because either one maybe
wrong as to cause or data loss or break transaction semantics.
Then it will mark the DTX entry as 'corrputed' globally. That
will prevent subsequent read operations against such DTX via
returning -DER_DATA_LOSS. On the other hand, the 'corrupted'
DTX only affects its modified pieces, but not affect others
that share the same target (object/key). The targets attached
to the 'corrupted' DTX will not be purged by vos aggregation,
but may be freshed by new update/punch. We need special tools
to handle (recover or cleanup) those 'corrupted' DTX entries
in the future.

The patch also fixes current confused 'rebuild' intent for
DTX availability check. We use new 'DAOS_INTENT_MIGRATION'
for all rebuild and migration related logic. The DTX logic
in VOS needs some special handling for the operations with
such intent, for example, return 'ALB_UNAVAILABLE' instead
of '-DER_DATA_LOSS' when hit 'corrupted' DTX entry.

It also contains some code cleanup for 'enum obj_rpc_flags'.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

File: src/include/daos_errno.h [hunk 1]
```diff
@@ -245,7 +245,10 @@ extern "C" {
 	       Agent is incompatible with libdaos)			\
 	/** Multiple shards locate on the same target */		\
 	ACTION(DER_SHARDS_OVERLAP,	(DER_ERR_DAOS_BASE + 30),	\
-	       Shards overlap)
+	       Shards overlap)						\
+	/** Not sure whether the TX is committable or not */		\
+	ACTION(DER_TX_UNCERTAINTY,	(DER_ERR_DAOS_BASE + 31),	\
```
Comment (NiuYawei) [inline src/include/daos_errno.h:250]: The term "uncertainty" is already used for checking time skew in distributed transaction, I think we'd use some different term here to avoid confusion.

File: src/include/daos_srv/vos_types.h [hunk 2]
```diff
@@ -240,6 +245,8 @@ enum {
 	VOS_OF_PUNCH_PROPAGATE		= (1 << 14),
 	/** replay punch (underwrite) */
 	VOS_OF_REPLAY_PC		= (1 << 15),
+	/** For detecting DTX uncertainty  */
+	VOS_OF_DETECT_UNCERTAINTY	= (1 << 16),
```
Comment (NiuYawei) [inline src/include/daos_srv/vos_types.h:249]: Same here, I suppose 'uncertainty' here means 'we don't know if the DTX is committed'? Should we use some other term for avoid confusion?

File: src/dtx/dtx_srv.c [hunk 3]
```diff
@@ -92,7 +94,31 @@ dtx_handler(crt_rpc_t *rpc)
 		else
 			rc = vos_dtx_check(cont->sc_hdl,
 					   din->di_dtx_array.ca_arrays,
-					   NULL, NULL, false);
+					   NULL, NULL, NULL, false);
+		break;
+	case DTX_REFRESH:
```
Comment (NiuYawei) [inline src/dtx/dtx_srv.c:99]: Could you add some comments to explain "DTX_REFRESH"?

Comment (Nasf-Fan) [reply src/dtx/dtx_srv.c:99]: Because of async batched commit semantics, the DTX status on the leader maybe different from the one on non-leaders. For the leader, it exactly knows whether the DTX is committable or not, but the non-leader does not know if the DTX is in 'prepared' status. So someone needs to know whether the 'prepared' DTX is real committable or not, the non-leader need to refresh its DTX status from the leader. The DTX_REFRESH RPC is used for that. I will add comment in the patch.

File: src/include/daos_srv/vos_types.h (hunk 2 above)
Comment (Nasf-Fan) [reply src/include/daos_srv/vos_types.h:249]: Yes, it means "we do not know whether some DTX is committable or not". Similarly as above, I am not sure which one is better.

File: src/include/daos_errno.h (hunk 1 above)
Comment (Nasf-Fan) [reply src/include/daos_errno.h:250]: To avoid confusion, we add the prefix "_TX_", just as "DER_TX_RESTART" does. I am not against to use another term, but not sure which one is better. Do you have any suggested name?

File: src/include/daos_srv/vos_types.h (hunk 2 above)
Comment (Nasf-Fan) [reply src/include/daos_srv/vos_types.h:249]: I will remove this flag in my new patch.

File: src/common/btree.c [hunk 4]
```diff
@@ -1756,14 +1761,21 @@ dbtree_fetch(daos_handle_t toh, dbtree_probe_opc_t opc, uint32_t intent,
 		return -DER_NO_HDL;
 
 	rc = btr_probe_key(tcx, opc, intent, key);
-	if (rc == PROBE_RC_INPROGRESS) {
+	switch (rc) {
+	case PROBE_RC_INPROGRESS:
```
Comment (jolivier23) [inline src/common/btree.c:1765]: Is this even possible anymore?  The committed state of object and key trees is stored in the incarnation log, not the key/object tree.   A probe of the object or key tree will never hit uncommitted data.  Or is it a different tree?

File: src/include/daos_errno.h (hunk 1 above)
Comment (jolivier23) [reply src/include/daos_errno.h:268]: DER_TX_UNKNOWN ?  I agree that uncertain is confusing since we use DER_TX_RESTART if we hit an uncertainty violation

File: src/vos/tests/vts_mvcc.c [hunk 5]
```diff
@@ -1486,6 +1492,25 @@ uncertainty_check_exec_one(struct io_test_args *arg, int i, int j, bool empty,
 		print_message("FAILED: "DF_CASE"\n",
 			      DP_CASE(i, j, empty, w, wp, we, commit, a, ap, ae,
 				      bound, mvcc_arg->i));
+
+	if (!daos_is_zero_dti(&wtx->th_saved_xid)) {
+		if (wtx->th_skip_commit)
```
Comment (jolivier23) [inline src/vos/tests/vts_mvcc.c:1497]: if skip_commit, then commit ?

File: src/vos/tests/vts_mvcc.c [hunk 6]
```diff
@@ -1486,6 +1492,25 @@ uncertainty_check_exec_one(struct io_test_args *arg, int i, int j, bool empty,
...
+			vos_dtx_commit(arg->ctx.tc_co_hdl, &wtx->th_saved_xid,
+				       1, NULL);
+		else
+			vos_dtx_abort(arg->ctx.tc_co_hdl, DAOS_EPOCH_MAX,
+				      &wtx->th_saved_xid, 1);
+	}
+
+	if (!daos_is_zero_dti(&atx->th_saved_xid)) {
+		if (atx->th_skip_commit)
```
Comment (jolivier23) [inline src/vos/tests/vts_mvcc.c:1506]: same comment here

File: src/vos/tests/vts_mvcc.c [hunk 7]
```diff
@@ -1486,6 +1492,25 @@ uncertainty_check_exec_one(struct io_test_args *arg, int i, int j, bool empty,
...
+			vos_dtx_commit(arg->ctx.tc_co_hdl, &atx->th_saved_xid,
+				       1, NULL);
+		else
+			vos_dtx_abort(arg->ctx.tc_co_hdl, DAOS_EPOCH_MAX,
```
Comment (jolivier23) [inline src/vos/tests/vts_mvcc.c:1510]: did the semantics change?  I did abort before but it caused issues if the tx failed.

File: src/vos/vos_io.c [hunk 8]
```diff
@@ -1379,6 +1410,9 @@ akey_update_single(daos_handle_t toh, uint32_t pm_ver, daos_size_t rsize,
 	daos_epoch_t		 epoch = ioc->ic_epr.epr_hi;
 	int			 rc;
 
+	if (vos_detect_dtx_uncertainty())
```
Comment (jolivier23) [inline src/vos/vos_io.c:1413]: definitely need a better name

File: src/vos/vos_tls.h [hunk 9]
```diff
@@ -128,7 +128,12 @@ vos_dth_set(struct dtx_handle *dth)
 static inline struct dtx_handle *
 vos_dth_get(void)
 {
-	return vos_tls_get()->vtl_dth;
+	struct vos_tls	*tls = vos_tls_get();
```
Comment (jolivier23) [inline src/vos/vos_tls.h:131]: why is this needed?

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/common/btree.c (hunk 4 above)
Comment (Nasf-Fan) [reply src/common/btree.c:1765]: Here, the probe_key is not only for object/dkey/akey, but also can be for single value. If some update just refreshes the single value under an existing akey, then it is possible that some later dbtree_fetch() of such SV will hit DER_INPROGRESS.

File: src/vos/tests/vts_mvcc.c (hunk 5 above)
Comment (Nasf-Fan) [reply src/vos/tests/vts_mvcc.c:1497]: skip_commit means that we does not want to commit related DTX to generate some special test condition for the test. But after the test, it is OK to commit it since it is done successfully.

File: src/vos/tests/vts_mvcc.c (hunk 6 above)
Comment (Nasf-Fan) [reply src/vos/tests/vts_mvcc.c:1506]: Same as above.

File: src/vos/tests/vts_mvcc.c (hunk 7 above)
Comment (Nasf-Fan) [reply src/vos/tests/vts_mvcc.c:1510]: In my test, I found that some old DTX entries affect the subsequent tests, but I cannot exactly point out which test cases left them, then I make a generally cleanup before each test case exit to guarantee that next run will be in a clean environment.

File: src/vos/vos_tls.h (hunk 9 above)
Comment (Nasf-Fan) [reply src/vos/vos_tls.h:131]: If without that, we will hit NULL pointer trouble when run vos evtree test.

File: src/vos/vos_io.c (hunk 8 above)
Comment (Nasf-Fan) [reply src/vos/vos_io.c:1413]: I will consider a better and clear name for that ...

**Comment by Nasf-Fan** [comment]
Rebase the patch against the latest master branch. Drop DER_TX_UNCERTAINTY from the patch.

**Comment by Nasf-Fan** [comment]
Please help to review again. Thanks!

File: src/object/srv_obj.c [hunk 10]
```diff
@@ -931,6 +970,13 @@ obj_fetch_create_maps(crt_rpc_t *rpc, struct bio_desc *biod)
 	 * Will be freed in obj_rw_reply
 	 */
 	iods_nr = orw->orw_iod_array.oia_iod_nr;
+
+	/* Re-entry case. */
+	if (orwo->orw_maps.ca_count != 0) {
+		D_ASSERT(orwo->orw_maps.ca_count == iods_nr);
+		return 0;
```
Comment (liuxuezhao) [inline src/object/srv_obj.c:977]: same, for re-entry case, the details of maps (determined by bio details) must be same as last time fetch?
not sure, just confirm

**Review by liuxuezhao — APPROVED** [review]
Just a few questions to confirm, thx

File: src/include/daos/dtx.h [hunk 11]
```diff
@@ -48,11 +48,11 @@ enum dtx_grp_flags {
...
 
 enum dtx_mbs_flags {
-	/* The targets that are modified by the distributed transaction
-	 * are in the same single redundancy group.
+	/* The targets modified via the DTX belong to replicated object
+	 * within single redundancy group.
 	 */
-	DMF_MODIFY_SRDG			= (1 << 0),
-	/* The MDS contains the leader information, used for distributed
+	DMF_SRDG_REP			= (1 << 0),
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/include/daos/dtx.h:54]: It is only for single redundancy group based replicated object. The EC object will handled differently.

File: src/include/daos_srv/vos_types.h [hunk 12]
```diff
@@ -38,7 +38,12 @@ struct dtx_rsrvd_uint {
 };
 
 enum dtx_cos_flags {
-	DCF_SHARED	= (1 << 0),
+	DCF_SHARED		= (1 << 0),
+	/* Some DTX (such as for the distributed transaction across multiple
+	 * RDGs, or for EC object modification) need to be committed via DTX
+	 * RPC instead of piggyback via other dispatched update/punch RPC.
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/include/daos_srv/vos_types.h:44]: For EC object, the leader and non-leader may have different data copy. The next EC update request may be dispatched to another set to servers.

"EXP_CMT" means explicitly commit.

File: src/object/srv_obj.c [hunk 13]
```diff
@@ -178,12 +185,6 @@ obj_rw_reply(crt_rpc_t *rpc, int status, uint64_t epoch,
 			orwo->orw_maps.ca_count = 0;
 		}
 
-		if (ioc->ioc_coc) {
-			daos_csummer_free_ic(ioc->ioc_coc->sc_csummer,
-				&orwo->orw_iod_csums.ca_arrays);
-			orwo->orw_iod_csums.ca_count = 0;
-		}
-
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/object/srv_obj.c:186]: The buffer has already been released via:

                if (orwo->orw_iod_csums.ca_arrays != NULL) {
                        D_FREE(orwo->orw_iod_csums.ca_arrays);
                        orwo->orw_iod_csums.ca_count = 0;
                }

File: src/object/srv_obj.c [hunk 14]
```diff
@@ -552,6 +555,12 @@ obj_set_reply_nrs(crt_rpc_t *rpc, daos_handle_t ioh, d_sg_list_t *sgls)
 	if (nrs_count == 0)
 		return 0;
 
+	/* Re-entry case. */
+	if (orwo->orw_nrs.ca_count != 0) {
+		D_ASSERT(orwo->orw_nrs.ca_count == nrs_count);
+		return 0;
+	}
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/object/srv_obj.c:562]: It should be. When fetch re-entry, if someone changed the VOS level layout during current ULT refreshing related DTX, then there will be two cases:

1). The change is triggered by someone with newer epoch, such update will be invisible for current fetch. So not affect current fetch.

2) The change is triggered by someone with older epoch, according to our MVCC mechanism, such update will be refused and required to be restarted with newer epoch. So this cannot happen.

So the orw_data_sizes must be the same as previous when re-entry.

**Review by liuxuezhao — APPROVED** [review]

File: src/object/srv_obj.c (hunk 10 above)
Comment (Nasf-Fan) [reply src/object/srv_obj.c:977]: Similar as above.

**Comment by Nasf-Fan** [comment]
Rebase against the latest master after Jeff's patch landing, fix CI test failure about EC degrade fetch.

**Review by liuxuezhao — DISMISSED** [review]

**Comment by Nasf-Fan** [comment]
Resolve merge conflict and degrade fetch failure.

**Comment by Nasf-Fan** [comment]
Update the patch to resolve DAOS-6252 issue.

**Comment by Nasf-Fan** [comment]
@jolivier23 , @NiuYawei,  the issues you mentioned in former review have been resolved. Please help review again. Thanks!

**Comment by Nasf-Fan** [comment]
@xuezhao, sorry, the patch is refreshed again after your last approve to resolve some CU test failures. Please help to review again. Thanks!

**Review by liuxuezhao — APPROVED** [review]

File: src/include/daos_srv/dtx_srv.h [hunk 15]
```diff
@@ -156,6 +176,23 @@ enum dtx_status {
 	DTX_ST_PREPARED		= 1,
 	/** The DTX has been committed. */
 	DTX_ST_COMMITTED	= 2,
+	/** The DTX is committable, but not committed. */
+	DTX_ST_COMMITTABLE	= 3,
+	/** The DTX is corrupted, some participant RDG(s) may be lost. */
+	DTX_ST_CORRUPTED	= 4,
+	/** The DTX is in-resync, not sure its status. */
+	DTX_ST_UNCERTAIN	= 5,
```
Comment (jolivier23) [inline src/include/daos_srv/dtx_srv.h:184]: Probably a different term here too.

File: src/common/btree.c [hunk 16]
```diff
@@ -953,6 +953,9 @@ btr_check_availability(struct btr_context *tcx, struct btr_check_alb *alb)
 	if (rc == -DER_INPROGRESS) /* Unceration */
 		return PROBE_RC_INPROGRESS;
 
+	if (rc == -DER_DATA_LOSS)
```
Comment (jolivier23) [inline src/common/btree.c:956]: Not really this patch, but just noticed the that Uncertain is not spelled correctly.  Not sure why our tools haven't flagged that one.

File: src/vos/vos_internal.h [hunk 17]
```diff
@@ -1183,6 +1184,25 @@ vos_epc_punched(daos_epoch_t epc, uint16_t minor_epc,
...
+vos_dtx_continue_detect(int rc)
+{
+	struct dtx_handle	*dth = vos_dth_get();
+
+	/* Continue to detect other potential in-prepared DTX. */
+	return rc == -DER_INPROGRESS && dth != NULL &&
+		dth->dth_share_tbd_count > 0 &&
+		dth->dth_share_tbd_count < DTX_DETECT_MAX;
+}
+
```
Comment (jolivier23) [inline src/vos/vos_internal.h:1205]: the line below is why I don't like the term uncertain for this case.

**Review by jolivier23 — APPROVED** [review]
A few minor nits but nothing blocking

**Comment by jolivier23** [comment]
> @jolivier23 , @NiuYawei, the issues you mentioned in former review have been resolved. Please help review again. Thanks!

Note, when you force land updates, it makes it really hard to figure out what you changed...actually quite impossible because it loses the context.

File: src/include/daos_srv/dtx_srv.h (hunk 15 above)
Comment (Nasf-Fan) [reply src/include/daos_srv/dtx_srv.h:184]: Then "DTX_ST_RESYNC" ?

File: src/common/btree.c (hunk 16 above)
Comment (Nasf-Fan) [reply src/common/btree.c:956]: I will fix it when refresh this patch or in subsequent patch.

File: src/vos/vos_internal.h (hunk 17 above)
Comment (Nasf-Fan) [reply src/vos/vos_internal.h:1205]: Yes, I noticed that there is already some concept has used the term of "uncertain".

**Comment by Nasf-Fan** [comment]
> Not for this patch but since SV is the only one using it, can we move this logic up to SV callbacks?

I am not sure, because btree logic will abstract (or convert) the error# returned by btr_ops API. That is why we specially handled -DER_INPROGRESS inside the btree logic. For example, if btr_probe_key() return PROBE_RC_ERR, then original dbtree_fetch() will handle it as NONEXIST case. It may be fine (or expected) for most of callers, but it is difficult for SV caller to understand what real happened. If we adjust btree logic to return the real errno#, I am afraid it will affect the other use cases. Personally, I prefer to cleanup btree logic, but it needs to be in another separated patch.

File: src/vos/evtree.c [hunk 18]
```diff
@@ -2320,9 +2453,17 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 		}
 	}
 out:
+	if (rc == 0 && !d_list_empty(&data_loss_list))
+		rc = evt_data_loss_check(&data_loss_list, ent_array);
```
Comment (NiuYawei) [inline src/vos/evtree.c:2457]: If this "data loss" check is done here, it'll affect entry insert and remove as well. I suppose the intention here is to notify app there is an error when app tries to fetch the "corrupted" data, however, this "corrupted" extent definitely shouldn't impede itself being deleted or other entries being inserted.

It looks to me that a simpler way is to carry the "corrupt" flag in evt_entry, and check the flag to see if it's fetching "corrupted" data in evt_find() (after array sort).

File: src/vos/vos_iterator.c [hunk 19]
```diff
@@ -490,25 +488,21 @@ reset_anchors(vos_iter_type_t type, struct vos_iter_anchors *anchors)
 {
 	switch (type) {
 	case VOS_ITER_DKEY:
-		D_ASSERT(daos_anchor_is_eof(&anchors->ia_dkey));
```
Comment (NiuYawei) [inline src/vos/vos_iterator.c:493]: Why these asserts are removed?

File: src/vos/evtree.c (hunk 18 above)
Comment (Nasf-Fan) [reply src/vos/evtree.c:2457]: I think we can add a parameter "for_read" for evt_ent_array_fill(), then we can only check the bad DTX case for evt_find() case.

File: src/vos/vos_iterator.c (hunk 19 above)
Comment (Nasf-Fan) [reply src/vos/vos_iterator.c:493]: Normally, when we call reset_anchors(), the former akey level iteration should have reached the end. But here, we exist former akey level iteration may because we hit some 'prepared' DTX, and then because we want to detect more 'prepared' DTX to improve the efficiency when check with leader, then we try to iterator a bit ahead.

File: src/dtx/dtx_common.c [hunk 20]
```diff
@@ -967,12 +1057,15 @@ dtx_handle_resend(daos_handle_t coh,  struct dtx_id *dti,
...
 again:
-	rc = vos_dtx_check(coh, dti, epoch, pm_ver, true);
+	rc = vos_dtx_check(coh, dti, epoch, pm_ver, NULL, true);
 	switch (rc) {
 	case DTX_ST_PREPARED:
 		return 0;
 	case DTX_ST_COMMITTED:
+	case DTX_ST_COMMITTABLE:
 		return -DER_ALREADY;
+	case DTX_ST_CORRUPTED:
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/dtx/dtx_common.c:1067]: I think we can rename it as "DTX_ST_RD_LOSS" if there is other DTX corrupted case in the future, then they are distinguishable.

File: src/common/btree.c [hunk 21]
```diff
@@ -1588,8 +1590,11 @@ btr_probe(struct btr_context *tcx, dbtree_probe_opc_t probe_opc,
 	rc = PROBE_RC_OK;
  out:
 	tcx->tc_probe_rc = rc;
+	if (rc == -DER_DATA_LOSS)
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/common/btree.c:1593]: OK, will chance it when refresh the patch next time or in the subsequent patch.

File: src/vos/evtree.c (hunk 18 above)
Comment (Nasf-Fan) [reply src/vos/evtree.c:2457]: So if we have chance to refresh the patch, then I will update it; otherwise, it will be done in subsequent DTX patch.

File: src/include/daos_srv/dtx_srv.h (hunk 15 above)
Comment (Nasf-Fan) [reply src/include/daos_srv/dtx_srv.h:184]: I found the solution to avoid "UNCERTAIN" name. I will do that if have chance to refresh the patch; otherwise, it will be in subsequent DTX patch.

**Comment by Nasf-Fan** [comment]
All CI tests passed.
pre-build/checkpatch failed for some known master code style issues.

File: src/vos/evtree.c (hunk 18 above)
Comment (Nasf-Fan) [reply src/vos/evtree.c:2457]: Sorry, Just realized that for insert and punch cases, the vos_dtx_check_availability() will NOT return -DER_DATA_LOSS for SV/EV record that attached to corrupted DTX, Your concerned issue has been resolved already. @NiuYawei

File: src/include/daos_srv/dtx_srv.h (hunk 15 above)
Comment (Nasf-Fan) [reply src/include/daos_srv/dtx_srv.h:184]: It is done in the patch PR#3939. @jolivier23

File: src/vos/evtree.c (hunk 18 above)
Comment (NiuYawei) [reply src/vos/evtree.c:2457]: Ok, thanks, but seems evt_delete() is using INTENT_PURGE, I'm not sure if that is overlooked in the patch.
